assets/js/ConfiguredHeadwaysForm.jsx

```diff
@@ -0,0 +1,191 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import {
+  Formik, Form, Field, FieldArray,
+} from 'formik';
+import {
+  object, string, array, number, ref,
+} from 'yup';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { faEdit, faTimes } from '@fortawesome/free-solid-svg-icons';
+import { timePeriodConfig } from './mbta';
+import { configuredHeadwayType } from './types';
+
+const ConfiguredHeadwaysForm = React.memo(({
+  branches,
+  configuredHeadways,
+  setConfiguredHeadways,
+  readOnly,
+  timePeriods,
+}) => {
+  const formSchema = React.useMemo(() => object().shape({
+    branches: array().of(
+      object().shape(timePeriods.reduce((acc, curr) => ({
+        ...acc,
+        [curr.id]: object().shape({
+          range_low: number().required().positive(),
+          range_high: number().required().positive().moreThan(ref('range_low')),
+        }),
+      }), { id: string().required() })),
+    ),
+  }), [timePeriods]);
+
+  const initialFormValues = React.useMemo(() => ({
+    branches: branches.map((branch) => {
+      const config = configuredHeadways[branch.id] || {};
+      return timePeriods.reduce((acc, curr) => ({
+        ...acc,
+        [curr.id]: config[curr.id] || { range_low: '', range_high: '' },
+      }), { id: branch.id });
+    }),
+  }), [branches, configuredHeadways]);
+
+  const isEnabled = React.useMemo(() => branches.some(branch => !!configuredHeadways[branch.id]), [branches, configuredHeadways]); // eslint-disable-line max-len
+
+  const [inEditMode, setInEditMode] = React.useState(!isEnabled);
+
+  React.useEffect(() => {
+    setInEditMode(!isEnabled);
+  }, [branches, configuredHeadways, isEnabled]);
+
+  const handleSubmit = React.useCallback((values) => {
+    const newConfig = values.branches.reduce((branchAcc, branchCurr, index) => ({
+      ...branchAcc,
+      [branchCurr.id]: timePeriods.reduce((timePeriodAcc, timePeriodCurr) => ({
+        ...timePeriodAcc,
+        [timePeriodCurr.id]: values.branches[index][timePeriodCurr.id],
+      }), {}),
+    }), {});
+    setConfiguredHeadways(newConfig);
+  }, [setConfiguredHeadways]);
+
+  return (
+    <div className="mb-5">
+      <div>
+        <h5 className="viewer--configured-headways-form-header">Headways</h5>
+      </div>
+      <Formik
+        validationSchema={formSchema}
+        enableReinitialize
+        initialValues={initialFormValues}
+        onSubmit={handleSubmit}
+      >
+        {({
+          dirty, isValid, resetForm,
```

Comment:
I don't think I've seen this before where in the JSX there's a function. I know that JSX elements like `<div>` compile to a function call of some sort, but what does it mean for a function to be here? Does it get called every render? Is this something specific to `Formik`?

Reply:
Basically the `Formik` component is setup to handle the `children` prop being a function (essentially a [render prop](https://reactjs.org/docs/render-props.html) or a JSX Element, calling the provided function within its own [`render`](https://github.com/jaredpalmer/formik/blob/master/packages/formik/src/Formik.tsx#L997) function. This isn't a default behavior of react, but I've seen a few libraries take this approach to passing props (usually internal state) from a parent component to a child.

Reply:
The alternative is the `Formik` component's `component` prop, which is also supposed to be a function (taking the props passed down by `Formik` as an arugment) and returning JSX.

https://jaredpalmer.com/formik/docs/api/formik#component-reactcomponenttypeformikpropsvalues